File: extension-system/sdk-node/src/extension.ts
import { EXTENSION_METHODS } from "@mocu/extension-contracts";

import type { ExtensionExecuteParams } from "@mocu/extension-contracts";

import { LlmApi } from "./llm.js";
import { DecisionApi } from "./decision.js";
import { EmbeddingApi } from "./embedding.js";
import { ExtensionUiApi } from "./ui.js";
import { JsonRpcProtocolClient } from "./protocol-client.js";

import type {
  ExtensionCommandHandler,
  ExtensionConfig,
  MocuExtensionDefinition,
} from "./types.js";

/**
 * The minimal Mocu extension SDK. An extension just registers command
 * handlers and calls `start()`; Mocu invokes the requested command on demand
 * via `extension.execute`. Extensions may also call host AI APIs and request
 * user interaction in Mocu chat through the execution context.
 */
export class MocuExtension {
  private readonly protocol = new JsonRpcProtocolClient();
  private readonly commands = new Map<string, ExtensionCommandHandler>();

  /** Call the Mocu host LLM from inside a command. */
  public readonly llm: LlmApi;

  /** Ask typed probabilistic questions via the Jev decision model. */
  public readonly decision: DecisionApi;

  /** Create embedding vectors with Mocu's configured embedding model. */
  public readonly embedding: EmbeddingApi;

  public constructor(
    private readonly definition: MocuExtensionDefinition = {},
  ) {
    this.llm = new LlmApi(this.protocol);
    this.decision = new DecisionApi(this.protocol);
    this.embedding = new EmbeddingApi(this.protocol);

    for (const [command, handler] of Object.entries(definition.commands ?? {})) {
      this.commands.set(command, handler);
    }

    this.protocol.registerRequestHandler(
      EXTENSION_METHODS.execute,
      async (params) => this.executeCommand(params as ExtensionExecuteParams),
    );
  }

  public registerCommand(
    command: string,
    handler: ExtensionCommandHandler,
  ): void {
    const normalizedCommand = command.trim();

    if (!normalizedCommand) {
      throw new Error("Command name cannot be empty.");
    }

    if (this.commands.has(normalizedCommand)) {
      throw new Error(`Command is already registered: ${normalizedCommand}`);
    }

    this.commands.set(normalizedCommand, handler);
  }

  public start(): void {
    this.protocol.start();
  }

  /** Send a one-way JSON-RPC notification to the Mocu host. */
  public notify(method: string, params?: unknown): void {
    this.protocol.notify(method, params);
  }

  private async executeCommand(
    params: ExtensionExecuteParams,
  ): Promise<{ success: boolean; output?: unknown; error?: string }> {
    try {
      let output: unknown;
      const rawContext =
        params.context && typeof params.context === "object"
          ? (params.context as Record<string, unknown>)
          : {};
      const context = {
        ...rawContext,
        mocu: {
          ui: new ExtensionUiApi(
            this.protocol,
            params.command,
            rawContext,
          ),
        },
      };
      const executionParams = { ...params, context };

      if (this.definition.execute) {
        output = await this.definition.execute(executionParams);
      } else {
        const handler = this.commands.get(params.command);

        if (!handler) {
          return {
            success: false,
            error: `Unknown command: ${params.command}`,
          };
        }

        const config: ExtensionConfig =
          params.config && typeof params.config === "object"
            ? (params.config as ExtensionConfig)
            : {};

        output = await handler(
          params.input,
          context,
          config,
        );
      }

      return { success: true, output };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }
}

export const createExtension = (
  definition: MocuExtensionDefinition = {},
): MocuExtension => new MocuExtension(definition);

export default MocuExtension;

File: extension-system/sdk-python/mocu_extension_sdk/extension.py
from __future__ import annotations

from typing import Any

from .decision import DecisionApi
from .embedding import EmbeddingApi
from .llm import LlmApi
from .protocol import JsonRpcProtocol
from .types import CommandHandler


class MocuExtension:
    """
    Minimal Mocu extension. An extension registers command handlers and calls
    `run()`; Mocu invokes the requested command on demand via
    `extension.execute`. Extensions may also call host APIs from inside a
    command: `extension.llm.generate()`, `extension.decision.ask()` and
    `extension.embedding.embed()`. Handlers receive `(input, context, config)`
    where `config` holds the values the user filled in on the extension's
    card in the Extensions page (manifest `config` fields).
    """

    def __init__(self) -> None:
        self._protocol = JsonRpcProtocol()
        self._commands: dict[str, CommandHandler] = {}

        # Let extensions call the Mocu host LLM, Jev decision model and
        # embedding model.
        self.llm = LlmApi(self._protocol)
        self.decision = DecisionApi(self._protocol)
        self.embedding = EmbeddingApi(self._protocol)

        self._protocol.register_handler(
            "extension.execute",
            self._handle_execute,
        )

    def command(
        self,
        name: str,
    ):
        normalized_name = name.strip()

        if not normalized_name:
            raise ValueError("Command name cannot be empty.")

        def decorator(handler: CommandHandler) -> CommandHandler:
            self.register_command(normalized_name, handler)
            return handler

        return decorator

    def register_command(self, name: str, handler: CommandHandler) -> None:
        normalized_name = name.strip()

        if not normalized_name:
            raise ValueError("Command name cannot be empty.")

        if normalized_name in self._commands:
            raise ValueError(f"Command is already registered: {normalized_name}")

        self._commands[normalized_name] = handler

    def run(self) -> None:
        self._protocol.run()

    def _handle_execute(self, params: Any) -> dict[str, Any]:
        if not isinstance(params, dict):
            return {
                "success": False,
                "error": "Execute params must be an object.",
            }

        command = params.get("command")

        if not isinstance(command, str):
            return {
                "success": False,
                "error": "Command must be a string.",
            }

        handler = self._commands.get(command)

        if handler is None:
            return {
                "success": False,
                "error": f"Unknown command: {command}",
            }

        input_value = params.get("input")
        context = params.get("context", {})
        config = params.get("config", {})

        if not isinstance(context, dict):
            context = {}

        if not isinstance(config, dict):
            config = {}

        try:
            output = handler(input_value, context, config)

            return {
                "success": True,
                "output": output,
            }
        except Exception as error:
            return {
                "success": False,
                "error": str(error),
            }


def create_extension() -> MocuExtension:
    return MocuExtension()
